src/codec.rs

```diff
@@ -554,25 +552,42 @@ impl<T> WrapperTypeDecode for Box<T> {
 		//
 		//         Constructing a `Box` from `NonNull::dangling` is also always safe as long
 		//         as the underlying type is zero-sized.
-		let mut boxed: Box<MaybeUninit<T>> = unsafe { Box::from_raw(ptr) };
-
+		let mut boxed = unsafe { Box::from_raw(ptr) };
 		T::decode_into(input, &mut boxed)?;
 
 		// Decoding succeeded, so let's get rid of `MaybeUninit`.
 		//
 		// TODO: Use `Box::assume_init` once that's stable.
-		let ptr: *mut MaybeUninit<T> = Box::into_raw(boxed);
-		let ptr: *mut T = ptr.cast();
+		let ptr: *mut T = Box::into_raw(boxed).cast();
 
 		// SAFETY: `MaybeUninit` doesn't affect the memory layout, so casting the pointer back
 		//         into a `Box` is safe.
-		let boxed: Box<T> = unsafe { Box::from_raw(ptr) };
+		let boxed = unsafe { Box::from_raw(ptr) };
 
 		input.ascend_ref();
 		Ok(boxed)
 	}
 }
 
+impl<T: Decode> Decode for Box<[T]> {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		Ok(Vec::decode(input)?.into_boxed_slice())
+	}
+}
+
+impl Decode for Box<str> {
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		// Guaranteed to create a Vec with capacity == len
+		let vec = Vec::from(Box::<[u8]>::decode(input)?);
+		// Guaranteed not to reallocate the vec, only transmute to String
+		let str = String::from_utf8(vec).map_err(|_| "Invalid utf8 sequence")?;
+
+		// At this point we have String with capacity == len,
+		// therefore String::into_boxed_str will not reallocate
+		Ok(str.into_boxed_str())
```

Comment:
Add `debug_assert_eq!`s for the `capacity == len`; then you don't need to spell it out in the comments.

Reply:
done